### PR TITLE
Elasticsearch: Add 7.x version support

### LINKED ISSUE #15622: Elasticsearch 7.x support
### What Grafana version are you using?
6.0.0 beta 3
### What datasource are you using?
Elasticsearch 7.0.0 beta 1
### What OS are you running grafana on?
Ubuntu
### What did you do?
After upgrading a Elasticsearch cluster from 6.6 to 7.0 beta 1, I can no longer add new indices as seperate datasources to grafana, due to the following error:
image

However, current data sources that were created when the cluster was 6.6.0 now no longer work, all charts display the following error:
image

Some useful F12 sceenshots to help:
1. Response when using old data sources after upgrading to Elasticsearch 7.0:
image

2. Response when trying to create new data sources that are 7.0:
the _mapping Json is the following:
```
{"logging-2019.02.24":{"mappings":{"properties":{"@timestamp":{"type":"date"},"@version":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"Host-City":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"Host-Country-Code":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"Host-Env":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"Host-Type":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"HostName":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"Log-Event":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},,"host":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"port":{"type":"long"},"type":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}}}}}}
```
Which clearly shows the @timestamp field as being of type 'date', even though grafana complains that @timestamp field does not exist

### PR DESCRIPTION
**Pull request by alcidesv**
Add '70' as a known ES version to go backend server
Let the data-source configurator know how to interpret a mapping in ES
ES 7 accepts the max number of shards as part of a "preferences" object
Add documentation regarding ES7
Get number of hits in a way which is compatible with ES 7 (is this dead code?)

Closes #15622

Obsoletes PR #16597

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 1]
```diff
@@ -234,8 +234,12 @@ export class ElasticDatasource {
       ignore_unavailable: true,
       index: this.indexPattern.getIndexList(timeFrom, timeTo),
     };
-    if (this.esVersion >= 56) {
+    if (this.esVersion >= 56 && this.esVersion < 70) {
       queryHeader['max_concurrent_shard_requests'] = this.maxConcurrentShardRequests;
+    } else if (this.esVersion >= 70) {
+      queryHeader['preference'] = {
```
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/datasource.ts]: Out of curiosity, where did you find information about this `preference` property change?

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 2]
```diff
@@ -347,8 +354,16 @@ export class ElasticDatasource {
...
+
+          if (configuredEsVersion < 70) {
+            for (const typeName in mappings) {
+              const properties = mappings[typeName].properties;
+              getFieldsRecursively(properties);
+            }
+          } else {
+            // ES 7 finally removed types in indices, and it also removed
+            // them from the API
+            const properties = mappings.properties;
```
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/datasource.ts]: I think this comment can be removed and instead you can add a test for es7 version. There's an existing test named `When getting fields` that you should be able to copy and change to use es7 version.

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 3]
```diff
@@ -283,6 +287,9 @@ export class ElasticDatasource {
   }
 
   getFields(query) {
+    // XXX: code-smell, I don't really know how to access the `this` closure
+    // value
+    const configuredEsVersion = this.esVersion;
```
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/datasource.ts]: Think this is OK :+1: you can remove the comment

File: public/app/plugins/datasource/elasticsearch/elastic_response.ts [hunk 4]
```diff
@@ -310,11 +310,15 @@ export class ElasticResponse {
...
+    const hitsTotal = typeof hits.total === 'number' ? hits.total : hits.total.value; // <- Works with Elasticsearch 7.0+
+
     const series = {
       target: 'docs',
       type: 'docs',
       datapoints: [],
-      total: hits.total,
+      // XXX: `hits.total` is an object in ES 7, not a number
+      // ... leaving comment while trying to determine where this is called ...
+      total: hitsTotal,
```
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/elastic_response.ts]: You can remove this comment and add a test instead. There's a test named `Raw documents query` in elastic_response.test.ts that you can copy and change to use es7 instead.

**Review by marefr — CHANGES_REQUESTED** [review]
Have tested and review this now, see comments for details. This looks very promising and we should be able to get this into 6.2. Great work :+1: 

I did also pushed some backend fixes to resolve v7 alerting/backend issues since I investigated why that didn't work and by that changed some code to make it work.

Besides this, regarding catasource configuration (config_ctrl.ts) it needs an update:
* If es version >= 7 then default `Max concurrent Shard Requests` should be 5 according to https://www.elastic.co/guide/en/elasticsearch/reference/7.0/breaking-changes-7.0.html#_semantics_changed_for_literal_max_concurrent_shard_requests_literal

I will also push an update of Grafana devenv docker blocks and provisioning for es 6 and es 7 - This we use in development.

**Comment by marefr** [comment]
@alcidesv please let us know if you need help the changes requested.

**Comment by alcidesv** [comment]
I will give it a try later today @marefr  and let you know!

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 2 above)
Comment (alcidesv) [reply public/app/plugins/datasource/elasticsearch/datasource.ts]: Ok, here is my attempt at adding some tests for this:  b3b3371 . As far as I can tell, those new tests are working OK. However I get failed tests which _seem_ to be connected: 

```
(node:11169) UnhandledPromiseRejectionWarning: Error: expect(received).toEqual(expected)

Difference:

- Expected
+ Received

- Array [
-   "@timestamp",
-   "beat.name.raw",
-   "beat.name",
-   "beat.hostname",
-   "system.cpu.system",
-   "system.cpu.user",
-   "system.process.cpu.total",
-   "system.process.name",
- ]
+ Array []
(node:11169) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:11169) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:11169) UnhandledPromiseRejectionWarning: Error: expect(received).toEqual(expected)

Difference:

- Expected
+ Received

- Array [
-   "system.cpu.system",
-   "system.cpu.user",
-   "system.process.cpu.total",
- ]
+ Array []
(node:11169) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:11169) UnhandledPromiseRejectionWarning: Error: expect(received).toEqual(expected)

Difference:

- Expected
+ Received

- Array [
-   "@timestamp",
- ]
+ Array []

```

Unfortunately, I'm unfamiliar with the way the test harness works, and can't pinpoint the message above to precise locations in the test suite.

File: public/app/plugins/datasource/elasticsearch/elastic_response.ts (hunk 4 above)
Comment (alcidesv) [reply public/app/plugins/datasource/elasticsearch/elastic_response.ts]: Done, commit b3b337151b089544930c72942634f3a3c8f7123a

File: docs/sources/features/datasources/elasticsearch.md [hunk 5]
```diff
@@ -59,7 +59,7 @@ a time pattern for the index name or a wildcard.
 ### Elasticsearch version
 
 Be sure to specify your Elasticsearch version in the version selection dropdown. This is very important as there are differences how queries are composed.
-Currently the versions available is 2.x, 5.x, 5.6+ or 6.0+. 5.6+ means a version of 5.6 or less than 6.0. 6.0+ means a version of 6.0 or higher, 6.3.2 for example.
+Currently the versions available is 2.x, 5.x, 5.6+, 6.0+ or 7.0+. 5.6+ means a version of 5.6 or less than 6.0. 6.0+ means a version of 6.0 or higher, 6.3.2 for example, and 7.0+ means version 7.0 and higher (hopefully).
```
[the remark this answers is not in the document]
Comment (alcidesv) [reply docs/sources/features/datasources/elasticsearch.md]: Here are some edits: 5b30823f1f7a2b2379250193fffe1dbe401d654e

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 1 above)
Comment (alcidesv) [reply public/app/plugins/datasource/elasticsearch/datasource.ts]: Hmm, good catch. 

ES 7 will complain about the parameter in the old position, but my edit is only hiding it from the query validator, not doing what I thought it did. 
For that matter,  [it didn't have the expected effect before either, it just didn't complain](https://github.com/elastic/elasticsearch/pull/35938) .

The correct way to pass this parameter is via a [URL query parameter](https://github.com/elastic/elasticsearch/blob/687978b7d167d4c20a29861b2558b2749c349529/server/src/main/java/org/elasticsearch/rest/RestRequest.java#L110), [apparently](https://github.com/elastic/elasticsearch/blob/622fb7883b4e7c6de6dce2636108e334cca28cd3/server/src/main/java/org/elasticsearch/rest/action/search/RestMultiSearchAction.java#L110).

I will try to fix and test this tomorrow.

Comment (alcidesv) [reply public/app/plugins/datasource/elasticsearch/datasource.ts]: Ok, now it's here: fb8491cca5316620d4d9bbdfae75615596f9afcd

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 2 above)
Comment (marefr) [reply public/app/plugins/datasource/elasticsearch/datasource.ts]: There was actually a problem with the tests. ES version wasn't provided in a correct way. Have pushed fix.

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 1 above)
Comment (marefr) [reply public/app/plugins/datasource/elasticsearch/datasource.ts]: Thanks. So frustrating that you actually have to look at the code to understand how their API works.

Comment (marefr) [reply public/app/plugins/datasource/elasticsearch/datasource.ts]: Did some changes. Based on what I can see in the code the search_type is still valid to provide in header of _msearch. Simplified some code and fix max concurrent shard requests in backend.

**Comment by holiiveira** [comment]
I have encountered this warning in the response header:: `Warning: 299 Elasticsearch-7.0.0-b7e28a7 "Deprecated aggregation order key [_term] used, replaced by [_key]"`

**Comment by marefr** [comment]
@holiiveira from which query - that was fixed long time ago (#14175)

**Comment by holiiveira** [comment]
@marefr 

Request:
```
POST /grafana/api/datasources/proxy/23/_msearch HTTP/1.1
Host: elk01.xxxx.com.br
Connection: keep-alive
Content-Length: 565
Pragma: no-cache
Cache-Control: no-cache
Authorization: xxxxx
Accept: application/json, text/plain, */*
Origin: https://elk01.xxxxx.com.br
X-Grafana-Org-Id: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.103 Safari/537.36
Content-Type: application/json;charset=UTF-8
Referer: https://elk01.xxxxxx.com.br/grafana/d/4NycrIkWz/background?tab=queries&panelId=11&edit&fullscreen&orgId=1&from=now-15m&to=now&var-comp=mtao016
Accept-Encoding: gzip, deflate, br
Accept-Language: pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7
Cookie: grafana_session=c7e202942e4ab8357b7ffef74b00897c; _ga=GA1.3.468404670.1550496714; __utma=187451450.468404670.1550496714.1553888618.1553888618.1; __utmz=187451450.1553888618.1.1.utmcsr=google|utmccn=(organic)|utmcmd=organic|utmctr=(not%20provided)
```

Response:
```
HTTP/1.1 200 OK
Server: nginx/1.14.0 (Ubuntu)
Date: Wed, 24 Apr 2019 14:41:00 GMT
Content-Type: application/json; charset=UTF-8
Content-Length: 206
Connection: keep-alive
Content-Encoding: gzip
Warning: 299 Elasticsearch-7.0.0-b7e28a7 "Deprecated aggregation order key [_term] used, replaced by [_key]"
```

Payload:

```
{"search_type":"query_then_fetch","ignore_unavailable":true,"index":"auditbeat-*"}
{"size":0,"query":{"bool":{"filter":[{"range":{"@timestamp":{"gte":"1556115951069","lte":"1556116851069","format":"epoch_millis"}}},{"query_string":{"analyze_wildcard":true,"query":"*"}}]}},"aggs":{"3":{"terms":{"field":"processoStatus","size":10,"order":{"_term":"desc"},"min_doc_count":1},"aggs":{"2":{"date_histogram":{"interval":"10s","field":"@timestamp","min_doc_count":0,"extended_bounds":{"min":"1556115951069","max":"1556116851069"},"format":"epoch_millis"},"aggs":{}}}}}}
```

**Comment by marefr** [comment]
@holiiveira cannot reproduce that. What does your panel json looks like that creates your request?

**Comment by holiiveira** [comment]
@marefr 

Visualization: Singlestat

Query:

```
{
  "xhrStatus": "complete",
  "request": {
    "method": "POST",
    "url": "api/datasources/proxy/23/_msearch",
    "data": "{\"search_type\":\"query_then_fetch\",\"ignore_unavailable\":true,\"index\":\"auditbeat-*\"}\n{\"size\":0,\"query\":{\"bool\":{\"filter\":[{\"range\":{\"@timestamp\":{\"gte\":\"1556123363565\",\"lte\":\"1556124263565\",\"format\":\"epoch_millis\"}}},{\"query_string\":{\"analyze_wildcard\":true,\"query\":\"*\"}}]}},\"aggs\":{\"3\":{\"terms\":{\"field\":\"statusProcesso\",\"size\":10,\"order\":{\"_term\":\"desc\"},\"min_doc_count\":1},\"aggs\":{\"2\":{\"date_histogram\":{\"interval\":\"10s\",\"field\":\"@timestamp\",\"min_doc_count\":0,\"extended_bounds\":{\"min\":\"1556123363565\",\"max\":\"1556124263565\"},\"format\":\"epoch_millis\"},\"aggs\":{}}}}}}\n"
  },
  "response": {
    "took": 0,
    "responses": [
      {
        "took": 0,
        "timed_out": false,
        "_shards": {
          "total": 1,
          "successful": 1,
          "skipped": 0,
          "failed": 0
        },
        "hits": {
          "total": {
            "value": 0,
            "relation": "eq"
          },
          "max_score": null,
          "hits": []
        },
        "aggregations": {
          "3": {
            "doc_count_error_upper_bound": 0,
            "sum_other_doc_count": 0,
            "buckets": []
          }
        },
        "status": 200
      }
    ],
    "$$config": {
      "method": "POST",
      "url": "api/datasources/proxy/23/_msearch",
      "data": "{\"search_type\":\"query_then_fetch\",\"ignore_unavailable\":true,\"index\":\"auditbeat-*\"}\n{\"size\":0,\"query\":{\"bool\":{\"filter\":[{\"range\":{\"@timestamp\":{\"gte\":\"1556123363565\",\"lte\":\"1556124263565\",\"format\":\"epoch_millis\"}}},{\"query_string\":{\"analyze_wildcard\":true,\"query\":\"*\"}}]}},\"aggs\":{\"3\":{\"terms\":{\"field\":\"statusProcesso\",\"size\":10,\"order\":{\"_term\":\"desc\"},\"min_doc_count\":1},\"aggs\":{\"2\":{\"date_histogram\":{\"interval\":\"10s\",\"field\":\"@timestamp\",\"min_doc_count\":0,\"extended_bounds\":{\"min\":\"1556123363565\",\"max\":\"1556124263565\"},\"format\":\"epoch_millis\"},\"aggs\":{}}}}}}\n"
    }
  }
}
```

Dashboard JSON:

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 113,
  "links": [],
  "panels": [
    {
      "cacheTimeout": null,
      "colorBackground": false,
      "colorValue": false,
      "colors": [
        "#299c46",
        "rgba(237, 129, 40, 0.89)",
        "#d44a3a"
      ],
      "datasource": "auditbeat-*",
      "format": "none",
      "gauge": {
        "maxValue": 100,
        "minValue": 0,
        "show": false,
        "thresholdLabels": false,
        "thresholdMarkers": true
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 11,
      "interval": "",
      "links": [],
      "mappingType": 1,
      "mappingTypes": [
        {
          "name": "value to text",
          "value": 1
        },
        {
          "name": "range to text",
          "value": 2
        }
      ],
      "maxDataPoints": 100,
      "nullPointMode": "connected",
      "nullText": null,
      "postfix": "",
      "postfixFontSize": "50%",
      "prefix": "",
      "prefixFontSize": "50%",
      "rangeMaps": [
        {
          "from": "null",
          "text": "N/A",
          "to": "null"
        }
      ],
      "sparkline": {
        "fillColor": "rgba(31, 118, 189, 0.18)",
        "full": false,
        "lineColor": "rgb(31, 120, 193)",
        "show": false
      },
      "tableColumn": "",
      "targets": [
        {
          "bucketAggs": [
            {
              "fake": true,
              "field": "statusProcesso",
              "id": "3",
              "settings": {
                "min_doc_count": 1,
                "order": "desc",
                "orderBy": "_term",
                "size": "10"
              },
              "type": "terms"
            },
            {
              "field": "@timestamp",
              "id": "2",
              "settings": {
                "interval": "auto",
                "min_doc_count": 0,
                "trimEdges": 0
              },
              "type": "date_histogram"
            }
          ],
          "metrics": [
            {
              "field": "select field",
              "id": "1",
              "type": "count"
            }
          ],
          "refId": "A",
          "timeField": "@timestamp"
        }
      ],
      "thresholds": "",
      "timeFrom": null,
      "timeShift": null,
      "title": "Panel Title",
      "type": "singlestat",
      "valueFontSize": "80%",
      "valueMaps": [
        {
          "op": "=",
          "text": "N/A",
          "value": "null"
        }
      ],
      "valueName": "avg"
    }
  ],
  "schemaVersion": 18,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-15m",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "5s",
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ],
    "time_options": [
      "5m",
      "15m",
      "1h",
      "6h",
      "12h",
      "24h",
      "2d",
      "7d",
      "30d"
    ]
  },
  "timezone": "",
  "title": "Background",
  "uid": "4NycrIkWz",
  "version": 38
}
```

Apparently the warning only occurs when the query returns nothing.

**Comment by marefr** [comment]
@holiiveira I cannot reproduce using your dashboard on current master or in this branch so I'll merge this and should there be any issues we hope those are caught in the 6.2 beta release phase.

**Review by marefr — APPROVED** [review]

**Comment by marefr** [comment]
Thank you for contributing to Grafana!

**Pull request merged**